servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultAsyncContextProvider.java

```diff
@@ -73,47 +66,33 @@ public ContextMap captureContext() {
 
     @Override
     public Scope attachContext(ContextMap contextMap) {
+        ContextMap prev;
         final Thread currentThread = Thread.currentThread();
         if (currentThread instanceof ContextMapHolder) {
             final ContextMapHolder asyncContextMapHolder = (ContextMapHolder) currentThread;
-            ContextMap prev = asyncContextMapHolder.context();
+            prev = asyncContextMapHolder.context();
             asyncContextMapHolder.context(contextMap);
-            return () -> detachContext(contextMap, prev == null ? newContextMap() : prev);
+            if (prev == null) {
+                prev = newContextMap();
+            }
         } else {
-            return slowPathSetContext(contextMap);
+            prev = CONTEXT_THREAD_LOCAL.get();
+            CONTEXT_THREAD_LOCAL.set(contextMap);
         }
+        return prev instanceof Scope ? (Scope) prev : new ScopeImpl(prev);
     }
 
-    private static Scope slowPathSetContext(ContextMap contextMap) {
-        ContextMap prev = CONTEXT_THREAD_LOCAL.get();
-        CONTEXT_THREAD_LOCAL.set(contextMap);
-        return () -> detachContext(contextMap, prev);
-    }
-
-    private static void detachContext(ContextMap expectedContext, ContextMap toRestore) {
+    @Override
+    public void setContextMap(ContextMap contextMap) {
         final Thread currentThread = Thread.currentThread();
         if (currentThread instanceof ContextMapHolder) {
             final ContextMapHolder asyncContextMapHolder = (ContextMapHolder) currentThread;
-            ContextMap current = asyncContextMapHolder.context();
-            if (current != expectedContext) {
-                LOGGER.debug("Current context didn't match the expected context. current: {}, expected: {}",
```

Comment:
While this approach reduces allocations from 1 to 0 (compare to my proposal with `CapturedContext.attach/detach`), it does not allow to make this validation. We don't have it today in the latest release, so it should be a blocker for this change. I just remember that similar checks in grpc-java & otel were quite useful for our partners to identity when the context is not propagated correctly.

If we go forward with this PR, we can consider still paying allocation cost of a new `Scope` when debug logging is enabled for this class.

Reply:
Done. I revived the `AsyncContextMapThreadLocal` from the git history since we need to set the thread local now from `CopyOnWriteContextMap` as well, so look there for details.

Reply:
Just kidding: I kept it in here for now.